### ISSUE TITLE
KRB5: Doc for new Liberty feature -- Kerberos for Ldap

### ISSUE THREAD

**Issue by kristip17** [OpenLiberty/docs#3571]
This is for feature https://github.com/OpenLiberty/open-liberty/issues/12794

General info will be very similar to our updates to the pages from the tWAS docs here: https://github.ibm.com/websphere/twas-docs/issues/3828

## Page Update, LDAP User Registry 3.0
Updates to https://openliberty.io/docs/21.0.0.1/reference/feature/ldapRegistry-3.0.html

## Item 1:
Update `The bind distinguished name and password are specified in the ladpRegistry element.`
to (just noticed that ldapRegistry needs to be corrected).
To bind with simple authentication, the bind distinguished name and password are specified in the ldapRegistry element.

## Item 2:
Add a new section: GSSAPI/Kerberos bind authentication (this is somewhat similar to the content on [https://openliberty.io/docs/21.0.0.1/kerberos-authentication-jdbc.html](https://openliberty.io/docs/21.0.0.1/kerberos-authentication-jdbc.html)

subtopic 1:
**Global Kerberos configuration for Open Liberty**

Kerberos authentication to an Ldap server in Open Liberty builds on the Kerberos Login Module and Java Generic Security Service (JGSS) API that are provided by the JDK.

The optional kerberos element provides configuration options for any Open Liberty features that use Kerberos credentials. If you have any Kerberos-enabled features configured for your server, you can specify this element in your server.xml file, as shown in the following example:

```
<kerberos keytab="${server.config.dir}/security/krb5.keytab" configFile="${server.config.dir}/security/krb5.conf"/>
```

The keytab and configFile attributes are both optional. If you do not specify these attributes, Open Liberty resolves their values according to the following priority:

The default locations that are defined by the JDK, such as the {user.home}/krb5.keytab file or a location that is specified by JDK-defined system properties

The default locations that are defined by the operating system, such as the /etc/krb5.keytab file or a location that is specified by operating system-defined environment variables

On the Windows operating system, the default Kerberos config file name is krb5.ini. On other operating systems, the file name is krb5.conf.

Subtopic 2:
**Kerberos configuration for Ldap servers**
To configure Kerberos bind authentication for Ldap servers, you must set the LdapRegistry attribute `bindAuthMechanism` to `GSSAPI`. For more information about LdapRegistry attibutes, see [LdapRegistry configuration](https://openliberty.io/docs/21.0.0.2/reference/config/ldapRegistry.html). After the Kerberos global configuration is optionally defined, a Kerberos principal can be configured on an ldapRegistry element in the server.xml file, as shown in the following example:

```
<ldapRegistry id="LDAP" realm="SampleLdapADRealm" host="ldap_hostname" port="389" ignoreCase="true"  baseDN="DC=example,DC=com" bindAuthMechanism="GSSAPI" krb5Principal="user1@EXAMPLE.COM" krb5TicketCache="${server.config.dir}/security/krb5-user1.cc" ldapType="Custom"  />
```
The krb5PrincipalName is required. 
The krb5TicketCache attribute is optional. If this attribute is not specified, Open Liberty checks the default Kerberos credential cache location of the JDK or the operating system.

If both the `krb5TicketCache` file and the `keytab` file from the global Kerberos config element are configured, then the `krb5TichetCache` file is search first and then the `keytab` file.

**Comment by dmuelle** [comment]
Hi @kristip17 - is this update targeted for a particular Open Liberty release?

**Comment by kristip17** [comment]
@dmuelle -- Our goal is a beta for 21003 and GA 21004.

Is there a good way to mark this as "not ready for review" yet? I wanted to start working on it in bits instead of jamming it at the end.

**Comment by dmuelle** [comment]
@kristip17 - no problem! We'll leave it in the "unprioritized" column until it's ready. You can just comment here to let me know when it's ready to be worked on.

**Comment by kristip17** [comment]
@dmuelle  -- this is ready to start review.

**Comment by dmuelle** [comment]
Thanks @kristip17, moving it to our backlog and we will target to publish with the 21.0.0.4 docs release

**Comment by kristip17** [comment]
We may be bumping GA to 21.0.0.5 -- this feature is in beta right now for 21.0.0.4-beta.

**Comment by chirp1** [comment]
OK with move of issue to 21.0.0.5 based on discussions with Kristi.

**Comment by dmuelle** [comment]
Content proposal:

Since we already document global KRB5 config on the Kerberos authentication for JDBC data sources page, propose moving the global info info to a new page in the Security section called "Kerberos authentication". This page will link out to related info, including the new LDAP example and the existing JDBC topic. I think the LDAP info should go with the other config examples on the LDAP  User registry feature page. That will keep all the LDAP info in the same place and avoid duplication of the Global info on multiple pages.

@kristip17 - this is a slightly different organization than what you propose above but I think it covers all the bases. With the update moving out to 21.0.0.5, do you expect any changes in the documentation content?

**Comment by kristip17** [comment]
@dmuelle -- a separate global info page on the `<kerberos/>` config that we can reference from the other articles sounds great. I hope to provide some additional troubleshooting information, but I think that content would be contained and either a new section or a new page.

**Comment by dmuelle** [comment]
Hi @kristip17 

I have a couple drafts for this issue ready for your review:

- [Kerberos authentication with Open Liberty](https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.4/kerberos-authentication.html)
This is the doc for the global `kerberos` element that other pages can link to. I've already updated the JDBC and LDAP docs on draft to point to it. I also revised the variable precedence info based on our slack conversation so let me know if I got that right.

- [Configure Kereberos authentication for LDAP servers](https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.4/reference/feature/ldapRegistry-3.0.html#krb5)
Added this info to the LDAP feature config page and also made the changes under "item 1" from your previous comment

Let me know if any further edits are needed, thanks!

**Comment by kristip17** [comment]
@dmuelle  -- The introductory sentences here seem awkward to me -- but we have the same in the DB2 Auth. Should we swap it around a bit? I don't have a strong feeling if you want to leave it as-is.

```
You can specify a global Kerberos configuration in your server.xml file that provides values to any Open Liberty features that use Kerberos credentials. This global configuration is optional. For more information, see Kerberos authentication for Open Liberty.

To configure Kerberos bind authentication for LDAP servers, you must configure the bind authentication mechanism and the Kerberos principal on the ldapRegistry element, as shown in the following example:
```

Maybe swap to

```
To configure Kerberos bind authentication for LDAP servers, you must configure the bind authentication mechanism and the Kerberos principal on the ldapRegistry element, as shown in the following example:

_Example_

In addition to the ldapRegistry configuration, you can specify a global Kerberos configuration in your server.xml file that provides values to any Open Liberty features that use Kerberos credentials. This global configuration is optional. For more information, see Kerberos authentication for Open Liberty.
```

**Comment by kristip17** [comment]
Also, I feel like I got weirdly repetitive on the last two paragraphs trying to describe ticketCache vs. keytab. I tried to mash them into 1 paragraph:

The krb5TicketCache attribute is optional and specifies the location of a ticket or credential cache (often called ccache) file. If this attribute is not specified, Open Liberty resolves credential values from the globally configured the keytab file that is configured in the global Kerberos element. If no keytab file is configured, Open Liberty resolves credential values from the default Kerberos credential cache location for the Java SDK or the operating system. If the ldapRegistry krb5TicketCache attribute and the Kerberos keytab attribute are both configured, Open Liberty searches for credentials first in the krb5TicketCache file and then in the keytab file. For more information on the global Kerberos element, see Kerberos authentication for Open Liberty.

I also think we should add this information (did not have in my original doc):

When the krb5TicketCache attribute is used, once the principal name is successfully authenticated, the Kerberos service will try to automatically renew the ticket before it expires.

**Comment by dmuelle** [comment]
>When the krb5TicketCache attribute is used, once the principal name is successfully authenticated, the Kerberos service will try to automatically renew the ticket before it expires.

Does this mean that the using attribute sets it up so that the service makes auto-renew attempts on the ticket that are not made if the config is set globally?



Also, we haven't clearly defined "ticket" in this doc. It says in your comment above 

>a ticket or credential cache (often called ccache) file...

is the ticket the same thing as the ccache file? Or an alternative option?

**Comment by dmuelle** [comment]
slack comments from @kristip17 

Hey David -- Does this mean that the using attribute sets it up so that the service makes auto-renew attempts on the ticket that are not made if the config is set globally?  Yes -- the keytab file doesn't contain credentials that expire. The krb5TicketCache contains credentials that expire and, depending on how the credential and LDAP server is configured, can be renewed. So you might add your credential to your krb5TicketCache file and it has a 24hr expiration. You start Liberty, we use that credential from the ticketCache and login into the Ldap server and things are happy. When Liberty tries to use that credential near or after the 24 hour expiration, we'll attempt to renew it (there's usually a max renewal time and the LDAP or KDC server has to allow renewals). If your credential was already expired when you start the Liberty server, it will just fail (we can't renew it then). Otherwise, a user has to refresh/renew the credential themselves (using standalone tools like kinit).
9:56
This a ticket or credential cache (often called ccache) file... could also be "a ticket cache or credential cache" -- basically krb5Ticketcache==ticket cache==credential cache==ccache. In tWAS, they seemed to be sort of loosey goosey about using ticketCache or credential cache or ccache. So now I'm not sure if it's more or less confusing to use all terms to reference it. Not really sure why we didn't just stick with credential cache/ccache -- https://web.mit.edu/kerberos/krb5-1.12/doc/basic/ccache_def.html (edited)

**Comment by dmuelle** [comment]
Hi @kristip17 - I've updated the draft per your comments, let me know what else it needs!

https://draft-openlibertyio.mybluemix.net/docs/latest/reference/feature/ldapRegistry-3.0.html#krb5

Thanks!

**Comment by kristip17** [comment]
On this, `...Open Liberty searches for credentials first in the ccache file and then in the keytab file.`  I'd either change ccache to krb5TicketCache or maybe tweak it like something like this:
`Open Liberty searches for credentials first in the ccache file defined by the krb5TicketCache attribute and then in the keytab file.`

**Comment by dmuelle** [comment]
Hi @kristip17 , draft is updated:

https://draft-openlibertyio.mybluemix.net/docs/21.0.0.4/reference/feature/ldapRegistry-3.0.html#krb5

Let me know if further changes are needed, if not you can sign off by adding the `technical reviewed` label to this issue. Thanks!

**Comment by kristip17** [comment]
I wonder if we should update `You can specify a global Kerberos configuration in your server.xml file that provides values to any Open Liberty features that use Kerberos credentials.` to include what you can add from there.

You can specify additional global Kerberos configuration (a keytab file and a configuration file) in your server.xml file that provides values to any Open Liberty features that use Kerberos credentials.

**Comment by dmuelle** [comment]
thanks @kristip17 - updated the draft to add that info;

https://draft-openlibertyio.mybluemix.net/docs/latest/reference/feature/ldapRegistry-3.0.html#krb5

**Comment by kristip17** [comment]
After updating the Beta blog post, we have some additional changes for the doc

Ut wanted to move away from using "global" for the Kerberos config.

From
You can specify a global Kerberos configuration in your server.xml file 
to
You can specify additional Kerberos configuration in your server.xml file 

From
This global configuration is optional. 
to
The Kerberos element configuration is optional. 
(or maybe remove this sentence altogether?)

From
Open Liberty resolves credential values from the keytab file that is configured in the global kerberos element. 
to
Open Liberty resolves credential values from the Kerberos keytab file that is configured in the `kerberos` element. 

From
Open Liberty resolves credential values from the default Kerberos credential cache location for the Java SDK or the operating system. 
to
Open Liberty resolves credential values from the default Kerberos credential cache location by using the Java SDK default locations.

**Comment by dmuelle** [comment]
Hi @kristip17 - I've made the edits from your previous comment:

https://draft-openlibertyio.mybluemix.net/docs/latest/reference/feature/ldapRegistry-3.0.html#krb5

 One thing I wanted to check- re

"Open Liberty resolves credential values from the default Kerberos credential cache location by using the Java SDK default locations."

So we don't want to mention operating system defaults in this context? On the [main OL kerberos page](https://draft-openlibertyio.mybluemix.net/docs/21.0.0.5/kerberos-authentication.html#_kerberos_configuration_for_open_liberty) it says that the OS defaults are the last priority, after JDK defaults (option 3). Is that still correct?

Also edited this statement slightly to be more specific- let me know if I got it right:

"If no `keytab` file or `krb5TicketCache` attribute is configured, Open Liberty resolves credential values by using the default Kerberos credential cache location for the Java SDK."

**Comment by ManasiGandhi** [comment]
# Content Review

@dmuelle The [Configure Kerberos authentication for LDAP servers ](https://draft-openlibertyio.mybluemix.net/docs/21.0.0.5/reference/feature/ldapRegistry-3.0.html#krb5) looks good! Just a few comments:

- [x] 	Typo: Configure Kereberos authentication for LDAP servers, change Kereberos to Kerberos.

- [x] Could you rephrase this sentence "The krb5TicketCache attribute is optional and specifies the location of a credential cache (ccache) file." to avoid the use of parentheses, for example, "The krb5TicketCache attribute is optional and specifies the location of the ccache file, which is a credential cache file."

- [x] 	Can you elaborate more on “default Kerberos credential cache location for the Java SDK"?
- [x] Under "Customize search filters", in "Configure a secure connection with LDAPS
To enable a secure connection to an LDAP server for encrypted communication via TLS or SSL protocols, configure the ldapRegistry element to enable Lightweight Directory Access Protocol Over Secure Socket Links (LDAPS).", Acrolinx flagged the use of the Latin expression "via"

**Comment by dmuelle** [comment]
Thanks for reviewing @ManasiGandhi. I made updates for each of the items specified in your comment. This draft is on hold pending updates from the security team, and might be pushed out to 21.0.0.6

**Comment by Rwalls1** [comment]
# Peer review
The example looks good, I just have a few comments:
## Configure Kerberos authentication for LDAP servers

- I think you should change "The credentials in a ccache file are subject to expiration.” to "The credentials in a ccache file expire.” to be more concise.
- There seems to be an inconsistency in the description of how the Kerberos principal is specified. This sentence says that, “The Kerberos principal is specified by the required `krb5PrincipalName` attribute. However, the config example references the `krb5Principal` attribute, which is slightly different. I think this discrepancy should be addressed. 
- I don’t think the sentence that starts with the phrase, “You can specify a Kerberos configuration” adds much value as it doesn’t highlight the purpose of “specifying the `Kerberos element`”, which is what I think a reader would want to know. It is also already implied that you are configuring Kerberos when you mention specifying the `kerberos` element. So, I think you should change the sentence to something like, “You can specify the `kerberos` element in your `server.xml` file to configure Kerberos authentication for all features that use Kerberos credentials."

**Comment by dmuelle** [comment]
Thanks for reviewing @Rwalls1 - I made the following changes per your review:

- [x]  change "The credentials in a ccache file are subject to expiration.” to "The credentials in a ccache file expire.” to be more concise.
- [x] `krb5PrincipalName` ---> `krb5Principal`
- [x] Revised description of `kerberos` element

**Comment by dmuelle** [comment]
doc is published with 21005
